stats.pl

```diff
@@ -0,0 +1,111 @@
+#
+use lib ("$ENV{'PERLLIB'}/vendor_perl");
+use Net::WebSocket::Server;
+use utf8;
+use JSON;
+use threads;
+use threads::shared;
+use threads 'exit' => 'threads_only';
+
+require($ENV{'THEME_ROOT'} . "/stats-lib.pl");
+do "$root_directory/websockets-lib-funcs.pl";
+
+# Get port number
+my ($port) = @ARGV;
+
+# Check if user is admin
+if (!webmin_user_is_admin()) {
+    &remove_miniserv_websocket($port);
+    &error_stderr("Access denied to non-master administrators!");
+    exit(2);
+}
+
+# Clean up when socket is terminated
+$SIG{'ALRM'} = sub {
+    &remove_miniserv_websocket($port);
+    &error_stderr("Timeout waiting for connection");
+    exit(1);
+    };
+alarm(60);
+
+# Log successful connection
+&error_stderr("Listening on port $port");
+
+# Used variables
+my ($wsconn, $thread);
+my $stats_extract :shared = 0;
+my $stats_interval :shared = 1;
+my $stats_running :shared = 0;
+
+# Start WebSocket server
+Net::WebSocket::Server->new(
+    listen     => $port,
+    on_connect => sub {
+        my ($serv, $conn) = @_;
+        &error_stderr("WebSocket connection established");
+        if ($wsconn) {
+            &error_stderr("Unexpected another connection attempted to the same port.");
+            $wsconn->disconnect();
+            return;
+        }
+        $wsconn = $conn;
+        alarm(0);
+        
+        $conn->on(
+            handshake => sub {
+                # Is the key valid for this Webmin session?
+                my ($conn, $handshake) = @_;
+                my $key   = $handshake->req->fields->{'sec-websocket-key'};
+                my $dsess = &encode_base64($main::session_id);
+                $key   =~ s/\s//g;
+                $dsess =~ s/\s//g;
+                if (!$key || !$dsess || $key ne $dsess) {
+                    &error_stderr("Key $key does not match session ID $dsess");
+                    $conn->disconnect();
+                    }
+                },
+            utf8 => sub {
+                my ($conn, $msg) = @_;
+                utf8::encode($msg) if (utf8::is_utf8($msg));
+                my $data = decode_json($msg);
+                $stats_extract = $data->{'save'};
+                $stats_interval = $data->{'interval'};
+                $stats_running = $data->{'running'};
+                my $collect = sub {
+                    while ($stats_running) {
+                        $wsconn->send_utf8(encode_json(stats($stats_extract)));
+                        sleep($stats_interval);
+                    }
+                };
+
+                # Ensure thread is restarted if needed
+                if (!$stats_running) {
+                    $thread->join() if ($thread && $thread->is_joinable());
+                    $thread = undef;
+                }
+                # Initiate non-blocking data collection using shared
+                # variables for controlling the loop from the client
+                if ($stats_running && (!$thread || !$thread->is_running())) {
```

Comment:
Why is this thread needed?

Reply:
Without using threads, it wouldn't be possible to control `$collect` for other connections!

Reply:
Wait, so are we still supporting multiple concurrent clients using the same websockets server?

Reply:
> Wait, so are we still supporting multiple concurrent clients using the same websockets server?

Yes, we have to, as I [explained it in detail](https://github.com/webmin/authentic-theme/pull/1695#issuecomment-2144102377). But even if we didn't we would still had to support it because the same user could have multiple browser tabs opened.

Reply:
Multiple tabs would still share the same session ID. Does this patch still support multiple different users though?

Reply:
> Multiple tabs would still share the same session ID.

Yes, they would. However, this line of code isn't directly related to a user or a session. We don't care about the session. It can be the same user logged in from different computers, or it can be different users. What matters is that we require the client (i.e., browser) to send a hash of the session ID, which we later check on the server side. We verify if the user trying to interact with the socket is actually a _root_-capable user. If not, the user will be disconnected and won't receive broadcast updates of the stats.

> Does this patch still support multiple different users sharing a server though?

Yes, the current code works like this:

1. Any _root_-capable user logging into the system with the Authentic theme and having the "real-time monitoring" option enabled will activate the socket. If the socket is already activated, the user will just receive details on how to connect to the socket.
2. Later, any _root_-capable user, regardless of the session, will be able to collect stats from the socket, which will only have a single call to the `stats()` function making it much more efficient.

Reply:
I see, that makes sense.  But is all the code for that in this PR? I don't see in stats-socket.cgi how it returns an existing websockets port..

Reply:
> But is all the code for that in this PR? I don't see in stats-socket.cgi how it returns an existing websockets port

After it was merged and we agreed on making it simpler by dropping the old polling stats, I renamed it. So if you want to look at the code, it would be in the `stats.cgi` and `stats.pl` files.